### ISSUE TITLE
Keras classifiers?

### ISSUE THREAD

**Issue by besherh** [scikit-learn-contrib/DESlib#168]
Please note that this is not an issue!

I was wondering if I could use see support deep learning classifiers (keras)?

is this possible?

**Comment by Menelau** [comment]
Hello,

Yes it can support as long as the classifier implement the scikit-learn interface (predict, predict_proba).

**Comment by besherh** [comment]
Thanks for the reply,
I have found an example [Here](https://machinelearningmastery.com/use-keras-deep-learning-models-scikit-learn-python/) that shows how to implement scikit-learn interface.

Let's assume that I have an ensemble of 30 deep learning models (Keras models in h5 format), how to create my 'pool_classifiers' in order to pass it to one of the DSCs. 

Could you advise?
Thanks in advance!

**Comment by Menelau** [comment]
Hello,

In this case you just need to create a list of classifiers in order to pass it to the DS method:

`pool_classifiers = [model1, model2, ... model30]`

You can also check on this example on how to create a pool of classifiers manually just by adding them to a list: https://deslib.readthedocs.io/en/latest/auto_examples/example_heterogeneous.html#sphx-glr-auto-examples-example-heterogeneous-py

**Comment by besherh** [comment]
Thanks that was helpful (Y)

**Comment by besherh** [comment]
Hi again, 
I was looking for an ensemble pruning library, and I found: https://github.com/viisar/brew, however the project is dead now and it had been advised to check DESlib.
Are any ensemble pruning techniques that are currently supported in DESlib?

**Comment by Menelau** [comment]
@besherh 

Well DESlib has one static selection method which you can use to prune your ensemble: https://deslib.readthedocs.io/en/latest/modules/static/static_selection.html

It is a very simple ensemble pruning model.  As the main focus of the library is on dynamic ensembles, this part was not given much attention until now and there is still a lot of room for improvements in the static selection/pruning part of the DESlib.  For example, considering also the diversity measures in the pruning (which we already offer  in the utils module: https://deslib.readthedocs.io/en/latest/modules/util/diversity.html). 

If you have any ideas of an ensemble pruning method that you would like to see in the library  please let me know and I can see what I can do to add it. I don't know exactly which ensemble pruning algorithm was available on the brew library, however I will check that as soon as possible.

**Comment by besherh** [comment]
Thanks, I will check the static_selection (I could have questions for the future :) ), and you are right about considering the diversity measures as acc measures are not enough.
As you know, there are plenty of pruning techniques for ensembles, and it could mainly be divided into three categories: 
1- ordering-based pruning: like Kappa pruning & Reduce-error pruning.
2- clustering-based pruning: like k-means.
3- optimisation-based pruning: GA approaches.

If we could see one/two per each category, it will be more than great as we get the chance to compare between diffrent algorithms!

**Comment by besherh** [comment]
I am trying to fit the static_selection with cifar10 data, but there is an issue with the input dimensions:
First I am creating a static_selection object
`static_class = StaticSelection(models_pool)
`then trying to call fit on my input data : `static_class.fit(X_prune,y_prune)`. 
Error:
`Found array with dim 4. Estimator expected <= 2`
My input dimension:  
(40000, 32, 32, 3)
(40000, 10)

Any advice is going to be appreciated!
Thanks in advance

**Comment by Menelau** [comment]
@besherh 

Sorry for the late reply. I believe this error is due to the fact that the KerasClassifier (which you use to wrap your model) was not trained. In order to have the 'classes_' attribute, you need to actually train the returned model from KerasClassifier to make use of scikit-learn functionalities. 

This error happens in which line of code?

**Comment by besherh** [comment]
Hi,
The error is raised when calling static_prune.fit().
I have searched for the error and you are right.
I have to look about what I have to do regarding this, as a side note do you think transforming the input images from 3d to 2d (to work with this library) is safe ?

**Comment by Menelau** [comment]
It should not be a problem as long as you can reshape it back to format your model accepts as input.

However, thinking about that, I believe it is better to remove the restriction imposed in the static selection to accept only a 2d array as input. As it can be seen as a meta-estimator, it should not impose constraints on the input type, rather the estimator itself should deal with the checking and validating whether the input is in the correct standard. 

I will check if I can safely remove that constraint this weekend and update the static selection. methods.
